state-sync/storage-service/server/src/lib.rs

```diff
@@ -1077,53 +1100,94 @@ impl StorageReaderInterface for StorageReader {
         end_version: u64,
         include_events: bool,
     ) -> Result<TransactionListWithProof, Error> {
+        // Calculate the number of transaction outputs to fetch
         let expected_num_transactions = inclusive_range_len(start_version, end_version)?;
-        let max_transaction_chunk_size = self.config.max_transaction_chunk_size;
-        if expected_num_transactions > max_transaction_chunk_size {
-            return Err(Error::InvalidRequest(format!(
-                "Requested number of transactions is larger than the maximum! \
-             Requested: {:?}, maximum: {:?}.",
-                expected_num_transactions, max_transaction_chunk_size
-            )));
+        let max_num_transactions = self.config.max_transaction_chunk_size;
+        let mut num_transactions_to_fetch = min(expected_num_transactions, max_num_transactions);
+
+        // Attempt to serve the request
+        while num_transactions_to_fetch >= 1 {
+            let transaction_list_with_proof = self
+                .storage
+                .get_transactions(
+                    start_version,
+                    num_transactions_to_fetch,
+                    proof_version,
+                    include_events,
+                )
+                .map_err(|error| Error::StorageErrorEncountered(error.to_string()))?;
+
+            let (overflow_frame, num_bytes) = check_overflow_network_frame(
+                &transaction_list_with_proof,
+                self.config.max_network_chunk_bytes,
+            )?;
+            if !overflow_frame {
+                return Ok(transaction_list_with_proof);
+            } else {
+                increment_network_frame_overflow(
+                    DataResponse::TransactionsWithProof(transaction_list_with_proof).get_label(),
+                );
+                let new_num_transactions_to_fetch = num_transactions_to_fetch / 2;
+                debug!("The request for {:?} transactions was too large (num bytes: {:?}). Retrying with {:?}.",
+                    num_transactions_to_fetch, num_bytes, new_num_transactions_to_fetch);
+                num_transactions_to_fetch = new_num_transactions_to_fetch; // Try again with half the amount of data
+            }
         }
 
-        let transaction_list_with_proof = self
-            .storage
-            .get_transactions(
-                start_version,
-                expected_num_transactions,
-                proof_version,
-                include_events,
-            )
-            .map_err(|error| Error::StorageErrorEncountered(error.to_string()))?;
-        Ok(transaction_list_with_proof)
+        panic!(
```

Comment:
why panic? I feels better to return an Error instead

Reply:
I want to panic so we're immediately aware that there's an issue. Basically, if we ever hit this case, there's only two options: (i) someone misconfigured a node to use a really small frame size; or (ii) we have a bug in our and someone has found a way to send a message that exploits it. Either way, I don't want to deadlock and just keep spinning 😱

Reply:
I can imagine if there's one type of request that may fail with the frame size, let's say a single transaction with output is too big, but all other things still work fine

Reply:
@zekun000, do you think a single output will ever be bigger than 63.5MB? 😱

Reply:
I have a single frame limit 4mb and message size limit 32mb in my network pr, but no I don't expect it to ever be that huge, it's more like being conservative

Reply:
Agreed, this is why I want to panic in this case. If it's hit, it means something has gone horribly wrong 😄